core/src/Torch/Raw/Tensor/Generic.hs

```diff
@@ -199,53 +205,176 @@ getDynamicDim
 fillZeros :: (THTensorMath t, THTensor t, Num (HaskReal t)) => Ptr t -> IO (Ptr t)
 fillZeros t = c_fill t 0 >> pure t
 
+-- Display stuff
+minLogScale :: Int
+minLogScale =
+    ceiling $ logBase 10 p
+  where
+    minPositive = NL.minValue * NL.epsilon  -- get smallest denormal
+    p = if (minPositive == 0) then NL.minValue else minPositive
+
+numberFormat :: RealFloat a => Int -> Int -> [a] -> (String, a, Int)
+numberFormat precision minSz xs = (format, scale, sz)
+  where
+    listParams :: (RealFloat a) => [a] -> (a, a, Bool, Bool, Bool)
+    listParams z =
+      f (NL.infinity, -NL.infinity, True, False, False) z
+      where
+        f z [] = z
+        f (minAcc, maxAcc, intModeAcc, hasInvalidAcc, hasNegInfAcc) (x:xs) = 
+            f acc xs
+          where
+            val = abs x
+            newHasNegInf = (x == -NL.infinity) || hasNegInfAcc
+            acc =
+              if (isNaN val || val == NL.infinity) then
+                (minAcc, maxAcc, intModeAcc, True, newHasNegInf) 
+              else 
+                (newMin, newMax, newIntMode, newHasInvalid, newHasNegInf)
+                  where
+                    !newHasInvalid = hasInvalidAcc
+                    !newMin = min minAcc val
+                    !newMax = max maxAcc val
+                    !newIntMode = if (intModeAcc) then
+                                    if (val == NL.infinity) then intModeAcc
+                                    else mod' val 1 == 0
+                                  else 
+                                    False
+    prec = precision
+    defaultScale = 1.0
+    (minAbs, maxAbs, intMode, hasInvalid, hasNegInf) = listParams xs
+    expMin = if (minAbs > 0.0) then 1 + (floor $ logBase 10 minAbs) else 1
+    expMax = if (maxAbs > 0.0) then 1 + (floor $ logBase 10 maxAbs) else 1
+    myMinSz = if hasInvalid then max minSz $ 3 + (if hasNegInf then 1 else 0)
+              else max minSz 2
+    (format, scale, sz) =
+      let expSz = (prec + 4 + (max (length (show expMax)) 
+                                   (length (show expMin)))) in
+      if (intMode) then
```

Comment:
Looks like a good place for guards and case-statements. Also, this function is getting huge and should probably be split up into top-level helper functions (this way it can be documented and exported for reuse if someone else needs it).